File: iosApp/iosApp/service/impl/DashboardServiceImpl.swift
import Foundation
import AsyncHTTPClient

/**
 * DashboardService implementation that requests the dashboard from the Artemis server.
 */
class DashboardServiceImpl: DashboardService {

    let jsonProvider: JsonProvider

    init(jsonProvider: JsonProvider) {
        self.jsonProvider = jsonProvider
    }

    func loadDashboard(authorizationToken: String, serverUrl: String) async -> NetworkResponse<Dashboard> {
        let httpClient = HTTPClient(eventLoopGroupProvider: .createNew)

        let finalResponse = await performNetworkCall(httpClient: httpClient, maxBodyByteSize: 5 * 1024 * 1024, createRequest: {
            var request = HTTPClientRequest(url: serverUrl + "api/courses/for-dashboard")
            request.method = .GET
            request.headers.add(name: HttpHeaders.Accept, value: ContentTypes.Application.Json)
            request.headers.add(name: HttpHeaders.UserAgent, value: DefaultHttpHeaderValues.ArtemisUserAgent)

            request.headers.add(name: HttpHeaders.Authorization, value: "Bearer " + authorizationToken)
            return request
        }, decode: { body in
            try Dashboard(courses: jsonProvider.decoder.decode([Course].self, from: body))
        })

        try? httpClient.syncShutdown()

        return finalResponse
    }
}

File: androidApp/src/main/java/de/tum/informatics/www1/artemis/native_app/android/service/impl/courses/DashboardServiceImpl.kt
package de.tum.informatics.www1.artemis.native_app.android.service.impl.courses

import de.tum.informatics.www1.artemis.native_app.android.content.Course
import de.tum.informatics.www1.artemis.native_app.android.content.Dashboard
import de.tum.informatics.www1.artemis.native_app.android.service.DashboardService
import de.tum.informatics.www1.artemis.native_app.android.service.impl.KtorProvider
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.http.*

/**
 * [DashboardService] implementation that requests the dashboard from the Artemis server.
 */
class DashboardServiceImpl(
    private val ktorProvider: KtorProvider,
) : DashboardService {

    override suspend fun loadDashboard(
        authToken: String,
        serverUrl: String
    ): Dashboard {
        //Perform a network call to $serverUrl/api/courses/for-dashboard
        val courses: List<Course> = ktorProvider.ktorClient.get(serverUrl) {
            url {
                appendPathSegments("api", "courses", "for-dashboard")
            }
            bearerAuth(authToken)
        }.body() //Decode JSON to List<Course>

        return Dashboard(courses)
    }
}